File: python/grpc_server.py
import signal
import time
from concurrent import futures

import grpc

from proto.hello_pb2 import HelloInput, HelloOutput
from proto.hello_pb2_grpc import HelloServicer, add_HelloServicer_to_server


class HelloServicerImpl(HelloServicer):
    def __init__(self):
        super().__init__()

    def Hello(self, request, context):
        print('hello to', request.name)
        message = 'Hello! ' * request.repeat + request.name
        return HelloOutput(hello_message=message)


def sigterm_handler(signal, frame):
    global IS_ALL_READY
    _logger.warn("SIGTERM received, server is closing after 10 secs")
    IS_ALL_READY = False
    time.sleep(10)
    _logger.warn("server closed")
    import sys
    sys.exit(0)


if __name__ == '__main__':
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    hello_servicer = HelloServicerImpl()
    add_HelloServicer_to_server(hello_servicer, server)
    server.add_insecure_port('[::]:9988')
    server.start()
    print('Listening at port 9988')
    _ONE_DAY_IN_SECONDS = 60 * 60 * 24
    signal.signal(signal.SIGTERM, sigterm_handler)
    try:
        while True:
            time.sleep(_ONE_DAY_IN_SECONDS)
    except KeyboardInterrupt:
        server.stop(0)

File: cpp/grpc_server.cc
#include <iostream>
#include <unistd.h>

#include <grpc++/grpc++.h>
#include "proto/hello.grpc.pb.h"
#include "proto/hello.pb.h"

class HelloImpl final : public hello::Hello::Service{
  ::grpc::Status Hello(::grpc::ServerContext* context,
                       const ::hello::HelloInput* request,
                       ::hello::HelloOutput* response) override {
    std::cout << "hello to " << request->name() << std::endl;
    std::string message = "";
    for (int i = 0; i < request->repeat(); ++i) {
      message += "Hello! ";
    }
    message += request->name();
    response->set_hello_message(message);
    return ::grpc::Status::OK;
  }
};

void sigHandler(int signum) {
  std::cerr << "Terminating! Wait " << 10 << " seconds to finish processing...\n";
  sleep(10);
  std::cerr << "Terminated!";
  exit(signum);
}

int main() {
  std::string server_address = "[::]:9988";
  HelloImpl hello;

  ::grpc::ServerBuilder builder;
  builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
  builder.RegisterService(&hello);
  
  std::unique_ptr<::grpc::Server> server(builder.BuildAndStart());
  std::cout << "Server listening on " << server_address << std::endl;

  if (signal(SIGTERM, sigHandler) == SIG_ERR
    || signal(SIGINT, sigHandler) == SIG_ERR) {
    std::cerr << "handle signal failed";
    exit(1);
  }

  server->Wait();
}
